### PR TITLE
Make it immutable

### PR DESCRIPTION
**Pull request by kevva**
Fixes sindresorhus/chalk#63.

### CODE REVIEW

**Comment by kevva** [comment]
@sindresorhus ping.

**Comment by jbnicolai** [comment]
@kevva I'd rather fix this in a backwards compatible way. Perhaps the `module.exports` object both returning the styles, and a function to create a new instance?

**Comment by kevva** [comment]
Why do you need backwards compatibility?

**Comment by jbnicolai** [comment]
Existing projects will switch to the new major anyway, will have to figure out the changed API and implement a (small) change on their side. With [300.000](http://npm-stat.com/charts.html?package=ansi-styles) downloads a day, that easily adds up to a lot of wasted hours. Seeing how we can avoid this without much work, why not?

**Comment by kevva** [comment]
@jbnicolai, well, that's really not our problem. If they're upgrading a major they can't just expect it to be no API changes.

**Comment by sindresorhus** [comment]
I'm with @kevva on not caring for a major, but thinking about it I think it makes more sense to add a `create` method as most users won't need to create a fresh object and having the default be an object makes for a nicer interface. The `create` method should be non-enumerable.

**Comment by arthurvr** [comment]
> I think it makes more sense to add a create method as most users won't need to create a fresh object and having the default be an object makes for a nicer interface. The create method should be non-enumerable.

:+1: Sounds good to me.

**Comment by sindresorhus** [comment]
I'm still not sure about this...

Maybe it would make more sense to just move https://github.com/chalk/chalk/blob/f97382e56f96fbafcdbc328b529dbbe143aa9786/index.js#L16 here? That way all consumers of would benefit.

@jbnicolai @Qix- Thoughts?

**Comment by arthurvr** [comment]
> Maybe it would make more sense to just move https://github.com/chalk/chalk/blob/f97382e56f96fbafcdbc328b529dbbe143aa9786/index.js#L16 here? That way all consumers of would benefit.

I don't see why not. :+1:

**Comment by Qix-** [comment]
**tl;dr:** A legible blue is a feature of chalk on windows. It's specific to chalk and, being agnostic, a cloning feature shouldn't really be a part of this *ANSI style reference* library. Put a deep clone directly in chalk and then modify the blue value.

---

Well let's think about it logically. We have to identify what ansi-styles is being used for.

Is it a reference to *ANSI codes*, or is it a reference to *logical colors*? I would assume the former, simply because it should be a dictionary of *correct* codes (we can't assume this library is used solely for display purposes - Chalk just happens to use it for that purpose).

To be a *true* agnostic dependency, I personally believe it should remain a reference to *true* ANSI codes.

Why not, in chalk, simply clone the object coming from ansi-styles? It's a 4-liner, but it achieves what we want. It's Chalk specific functionality (chalk has the *feature* of having a legible blue on Windows - other dependents of ansi-styles may not want/expect that feature). As well, having a function that lives inside of ansi-styles and clones the object is a little... useless (again, thinking in terms of being an agnostic dependency).

**Comment by sindresorhus** [comment]
@Qix- Ok. Good point. I agree. The thing is. `ansi-styles` would need to be deep cloned in chalk, which would require another dependency, as `ansi-styles`. I'm trying not to introduce too many dependencies to chalk, since there's soo many dependents. So what do you think about the `.create` method on `ansi-styles` just for this purpose?

**Comment by Qix-** [comment]
@sindresorhus I guess so, though `create` is a misnomer I think. Why not `get`? We're not *creating* styles, we're *getting* them. I'm still skeptical (I still think it should be deep-cloned in chalk), but if we really want a function here then `get` or even `clone` seem more appropriate to me.

**Comment by sindresorhus** [comment]
Yeah, `get` sounds better. I don't like this either, but I'm trying to be pragmatic about it. I even tried the `lodash.cloneDeep()` method which says it supports something like structured clone, but it didn't clone `modifiers `/`.colors`/`bgColors` as those are created with `Object.defineProperty`. I'm not going to document the method, it's just for `chalk`. And we might come up with a better solution in the future. But I think this one is the most pragmatic one for now, unless anyone wants to convince me otherwise.

**Comment by Qix-** [comment]
@sindresorhus Have you thought about color spaces? `default` would be *true* ANSI colors, `win32

**Comment by sindresorhus** [comment]
It's not really `win32` though. It's more like [`win32` cmd.exe](https://github.com/chalk/chalk/blob/master/index.js#L15).

Hmm need to think about it.

**Comment by jbnicolai** [comment]
I'd rather go with the `create` method @sindresorhus suggested, @Qix-.

> Maybe it would make more sense to just move https://github.com/chalk/chalk/blob/f97382e56f96fbafcdbc328b529dbbe143aa9786/index.js#L16 here? That way all consumers of would benefit.

I'm strongly against this. Let's keep this library simple and free from any context-dependent behaviour, just a reflection of the spec, and let a consumer like chalk resolve any issues with the broken implementations out there.

Also, we've already introduced a second workaround in chalk for a windows specific issue, and I'd rather not add that here as well. See chalk/chalk#58 and chalk/chalk#72

File: index.js [hunk 1]
```diff
@@ -1,56 +1,60 @@
...
+		});
 	});
 
-	Object.defineProperty(styles, groupName, {
-		value: group,
-		enumerable: false
-	});
-});
+	// fix humans
+	styles.grey = styles.gray;
```
Comment (jbnicolai) [inline index.js]: Why did you add this here? Seems like an unrelated change - although I see why you did this, seeing how it's now immutable chalk can no longer monkeypatch.

Actually.. keep it, it's fine I guess ;)

**Comment by jbnicolai** [comment]
@kevva would you be willing to modify this PR to reflect the discussion above? I.e: return an object by default, but add a non-enumerable `create` property get a fresh copy.

**Comment by Qix-** [comment]
@sindresorhus @jbnicolai 

```javascript
Object.defineProperty(module, 'exports', {
  enumerable: true,
  get: createAnsiStyles;
});
```

Backwards compatible and non-breaking.

**Comment by sindresorhus** [comment]
@Qix- Yes! Awesome idea. Can you do a PR?

**Comment by kevva** [comment]
Sorry I never got to this, been away for some days. Seems like it's fixed now anyway :).

**Comment by jbnicolai** [comment]
@kevva no problem, thanks for getting things started :smile:
